server-common/src/main/java/org/apache/kafka/common/DirectoryId.java

```diff
@@ -121,8 +122,38 @@ public static Map<Integer, Uuid> createAssignmentMap(int[] replicas, Uuid[] dire
      * Create an array with the specified number of entries set to {@link #UNASSIGNED}.
      */
     public static Uuid[] unassignedArray(int length) {
+        return array(length, UNASSIGNED);
+    }
+
+    /**
+     * Create an array with the specified number of entries set to {@link #MIGRATING}.
+     */
+    public static Uuid[] migratingArray(int length) {
+        return array(length, MIGRATING);
+    }
+
+    /**
+     * Create an array with the specified number of entries set to the specified value.
+     */
+    private static Uuid[] array(int length, Uuid value) {
         Uuid[] array = new Uuid[length];
-        Arrays.fill(array, UNASSIGNED);
+        Arrays.fill(array, value);
         return array;
     }
+
+    /**
+     * Check if a directory is online, given a sorted list of online directories.
+     * @param dir              The directory to check
+     * @param sortedOnlineDirs The sorted list of online directories
+     * @return                 true if the directory is considered online, false otherwise
+     */
+    public static boolean isOnline(Uuid dir, List<Uuid> sortedOnlineDirs) {
+        if (UNASSIGNED.equals(dir) || MIGRATING.equals(dir)) {
```

Comment:
We should only allow for a directory ID to be MIGRATING if we are in migration mode.

Reply:
Can you clarify what you mean with "migration mode"?

If you mean zk->kraft migration, then I don't think so. Metadata records may have been persisted in an older version, without directory assignment. That may happen outside of a ZK->KRaft migration.
When such records are found, the default directory for every replica is assumed to be `MIGRATING`.

See in #14820 :
 * `new PartitionRegistration(PartitionRecord record)`
 * `partitionRegistration.merge(PartitionChangeRecord record)`
 * `partitionRegistration.toRecord(...)`
 * `partitionChangeBuilder.setAssignmentChanges()`

Reply:
> If you mean zk->kraft migration, then I don't think so. Metadata records may have been persisted in an older version, without directory assignment

yes, agreed. @pprovenzano , `MIGRATING` state can happen outside ZK migration.

Honestly, I think `MIGRATING` isn't a great name. Someone who sticks on metadata version 15 won't ever "migrate" to having directory IDs. Maybe calling this `UNKNOWN` would be more descriptive? But that's another discussion I guess :)

Reply:
I don't think it's a bad name, we just need to make it clear the migration refers to the record version, and not to the cluster migration. 

You might not remember, but [the name was your suggestion](https://lists.apache.org/thread/38sbj42ssoksjq2p488vdk1992dw3j1t). 😄

Reply:
Heh. :)

Yeah I'm on the fence. With good enough docs maybe the name is ok...